ingester/src/buffer_tree/partition.rs

```diff
@@ -410,8 +417,12 @@ impl PartitionData {
     ///
     /// All subsequent calls to [`Self::sort_key`] will return
     /// [`SortKeyState::Provided`]  with the `new`.
-    pub(crate) fn update_sort_key(&mut self, new: Option<SortKey>) {
-        self.sort_key = SortKeyState::Provided(new);
+    pub(crate) fn update_sort_key(
+        &mut self,
+        new_sort_key: Option<SortKey>,
+        new_sort_key_ids: Option<SortedColumnSet>,
+    ) {
+        self.sort_key = SortKeyState::Provided(new_sort_key, new_sort_key_ids);
```

Comment:
Might be worth a defensive check to be sure we don't corrupt if things are broken:

```suggestion
        assert!(new_sort_key.is_some() ^ new_sort_key_ids.is_some());
        
        self.sort_key = SortKeyState::Provided(new_sort_key, new_sort_key_ids);
```

Please update the call documentation:

```
/// The caller MUST provide exactly one of the arguments as a Some, and the other MUST be a None.
```

That's how it should work right? We should include this in other function docs.

Reply:
I will add the assert.

Regarding the comment `The caller MUST provide exactly one of the arguments as a Some, and the other MUST be a None.` It is not like that. Both `new_sort_key` and `new_sort_key_ids` must be Some.  

To avoid changing so much in the code, the `SortKeyState` will store both `sort_key` and `sort_key_ids`. This PR is to add `sort_key_ids` into it.

Reply:
Added the assert.
I also updated the doc but a bit different from your comment

Reply:
If both must be `Some`, why take two `Option`?

Reply:
> To avoid changing so much in the code, the SortKeyState will store both sort_key and sort_key_ids. This PR is to add sort_key_ids into it.

Yeah I think this is the correct thing to do 👌 

> @domodwyer is there a reason this originally took just an Option<SortKey> parameter?

I think it was a mistake on my part - @NGA-TRAN if it always must be `Some` (which I'm pretty sure it should be) then the signature should probably not have `Option` at all and just take the two types :+1:

Reply:
@domodwyer and @savage-engineer :
I think removing `Option` here will need to do in  different PR. Here is the reason:

The argument `new_sort_key_ids: Option<SortedColumnSet>` is passed down all the way from reading the `sort_key_ids` from the partition catalog. And due to the needs of the migration process, this catalog `sort_key_ids` field was defined as NULL at first  and, as the result, it needs to return as Option<SortedColumnSet> in the `Partition` struct. Even though this catalog field it is NOT NULL now after last week migration, we need a different PR to make it  `SortedColumnSet` in the `Partition` struct instead of option first. This change will affect a lot of other code changes in all components, not only the Ingester.

So the proposal is we keep it as Option here and we check if of `Some` and in a future PR, where we replace partition's Option<SortedColumnSet> with SortedColumnSet, the option will be automatically removed here.

I know this is a bit pain but this is how things work for this migration so far. So many small steps to ensure things work correctly and the code changes for each PR are not too much

Reply:
Another option is I will remove the option here and `set_partition_sort_key` but I need to do `Some` check at a different place. Let me approach with that

Reply:
I also found another reason we do not want to remove Option here yet and should do it in a different PR.
If you look at the SortKeyState https://github.com/influxdata/influxdb_iox/blob/7a6de3d422e45db3d3cd7974d37e7e79134dbb48/ingester/src/buffer_tree/partition.rs#L30, they are all Options. If we remove Option here, we should remove option there, too and will lead to a lot of other code changes. I think let us keep them Option here and in a future PR in which the purpose is to remove Option only and everywhere, it will be a lot clearer

Reply:
So I ended up remove Options here and in set_partition_sort_key, too.
What I do not do is to remove Options from `SortKeyState` which will be done in a separate PR as stated above